webSocket/transceiver/transceiver.go

```diff
@@ -0,0 +1,230 @@
+	if len(payload) == 0 {
+		wt.log.Debug("wt.verifyPayloadAndSendAckIfNeeded(): empty payload")
+		return
+	}
+
+	payloadData, err := wt.payloadParser.ExtractPayloadData(payload)
+	if err != nil {
+		wt.log.Warn("wt.verifyPayloadAndSendAckIfNeeded: cannot extract payload data", "error", err.Error())
+		return
+	}
+
+	err = wt.payloadHandler.ProcessPayload(payloadData.Payload)
+	if err != nil && wt.blockingAckOnError {
+		wt.log.Debug("wt.payloadHandler.ProcessPayload: cannot handler payload", "error", err)
+		return
+	}
+
+	wt.sendAckIfNeeded(connection, payloadData)
+}
+
+func (wt *wsTransceiver) sendAckIfNeeded(connection webSocket.WSConClient, payloadData *data.PayloadData) {
+	if !payloadData.WithAcknowledge {
+		return
+	}
+
+	timer := time.NewTimer(wt.retryDuration)
+	defer timer.Stop()
+
+	counterBytes := wt.payloadParser.EncodeUint64(payloadData.Counter)
+	for {
+		timer.Reset(wt.retryDuration)
+
+		err := connection.WriteMessage(websocket.BinaryMessage, counterBytes)
+		if err == nil {
+			return
+		}
+
+		if !strings.Contains(err.Error(), data.ErrConnectionNotOpen.Error()) {
+			wt.log.Error("could not write acknowledge message", "error", err.Error(), "retrying in", wt.retryDuration)
+		}
+
+		wt.log.Debug("wt.sendAckIfNeeded(): cannot write ack", "error", err)
+
+		select {
+		case <-timer.C:
+		case <-wt.safeCloser.ChanClose():
+			return
+		}
+	}
+}
+
+// Send will prepare and send the provided WsSendArgs
+func (wt *wsTransceiver) Send(args data.WsSendArgs, connection webSocket.WSConClient) error {
+	assignedCounter := atomic.AddUint64(&wt.counter, 1)
+	newPayload := wt.payloadParser.ConstructPayloadData(args, assignedCounter, wt.withAcknowledge)
+
+	return wt.sendPayload(newPayload, assignedCounter, connection)
+}
+
+func (wt *wsTransceiver) sendPayload(payload []byte, assignedCounter uint64, connection webSocket.WSConClient) error {
+	errSend := connection.WriteMessage(websocket.BinaryMessage, payload)
+	if errSend != nil {
+		return errSend
+	}
+
+	if !wt.withAcknowledge {
+		return nil
+	}
+
+	return wt.waitForAck(assignedCounter, connection)
+}
+
+func (wt *wsTransceiver) waitForAck(assignedCounter uint64, connection webSocket.WSConClient) error {
+	timer := time.NewTimer(wt.retryDuration)
+	defer timer.Stop()
+
+	for {
+		mType, message, err := connection.ReadMessage()
+		if err != nil {
+			wt.log.Debug("s.waitForAck(): cannot read message", "id", connection.GetID(), "error", err)
+		}
+
+		if mType != websocket.BinaryMessage {
+			wt.log.Debug("received message is not binary message", "id", connection.GetID(), "message type", mType)
+			continue
+		}
+
+		wt.log.Trace("received ack", "remote addr", connection.GetID(), "message", message)
+
+		receivedCounter, err := wt.payloadParser.DecodeUint64(message)
+		if err != nil {
+			wt.log.Warn("cannot decode counter: bytes to uint64",
+				"id", connection.GetID(),
+				"counter bytes", message,
+				"error", err,
+			)
```

Comment:
missing continue here?

Reply:
added

Reply:
No, It's not ok as for the first counter it will think that the received counter matches the assigned counter.

Reply:
The counter will start from 1.

Reply:
refactored.